orbit/pkg/update/update.go

```diff
@@ -106,19 +105,25 @@ func NewUpdater(opt Options) (*Updater, error) {
 	}
 
 	tufClient := client.NewClient(opt.LocalStore, remoteStore)
-	var rootKeys []*data.PublicKey
-	if err := json.Unmarshal([]byte(opt.RootKeys), &rootKeys); err != nil {
-		return nil, fmt.Errorf("unmarshal root keys: %w", err)
-	}
 
+	// First try initializing via stored local metadata. Older Orbit packages were distributed with
+	// only the root *key* metadata and not the full root metadata (see
+	// https://github.com/theupdateframework/go-tuf/issues/379), so it's important that we try using
+	// the fully loaded root metadata in the local store before we try using what is provided via
+	// configuration.
 	meta, err := opt.LocalStore.GetMeta()
-	if err != nil || meta["root.json"] == nil {
-		var rootKeys []*data.PublicKey
-		if err := json.Unmarshal([]byte(opt.RootKeys), &rootKeys); err != nil {
-			return nil, fmt.Errorf("unmarshal root keys: %w", err)
+	if err == nil && meta["root.json"] != nil {
+		if err := tufClient.Init(meta["root.json"]); err != nil {
+			return nil, fmt.Errorf("client init with local root metadata: %w", err)
 		}
-		if err := tufClient.Init(rootKeys, 1); err != nil {
-			return nil, fmt.Errorf("init tuf client: %w", err)
+	} else {
+		// If that didn't work (because the metadata doesn't exist due to this being the first
```

Comment:
> first init

What's the "first init"? AFAICS packages generated by `fleetctl package` contain a `tuf-metadata.json`.
Is the "first init" the `fleetctl package` step?

Reply:
Ah, maybe I misunderstood how things work. I remembered `fleetctl updates roots` returns only the root key metadata, and then everything else is bootstrapped from that. Maybe the only place where we have a possible issue is during package generation? In that case, `fleetctl` will now need the full `root.json` rather than just the key metadata?

Just took a look at `orbit --help` and I see that there is no config for retrieving that root key metadata. I guess it's only used in `fleetctl package`

```
   --update-roots value                       Root key JSON metadata for update server (from fleetctl updates roots)
```

Is that your understanding?

In that case, maybe the only necessary changes are in `fleetctl updates roots` and `fleetctl package`?

Reply:
Ah, here's what I found out:

1. `fleetctl package`

`update.NewUpdater` uses what's provided via `fleetctl package --update-roots ...`, or harcoded keys in [defaultRootKeys](https://github.com/fleetdm/fleet/blob/289dc97ec5fbe4a9568f2a1ce69556992392176f/orbit/pkg/update/update.go#L34) for https://tuf.fleetctl.com, to initialize `tufClient`, then runs:
```
updater, err := update.NewUpdater(updateOpt)
if err != nil {
	return nil, fmt.Errorf("failed to init updater: %w", err)
}
if err := updater.UpdateMetadata(); err != nil {
	return nil, fmt.Errorf("failed to update metadata: %w", err)
}
```
Thus, the generated package always has a `tuf-metadata.json`.

2. `orbit` 

Will run the same (`update.NewUpdater`) but will always have a `tuf-metadata.json`. One edge case is that if one deletes `/opt/orbit/tuf-metadata.json` then orbit will attempt to use the hardcoded keys for https://tuf.fleetctl.com and sync the metadata again.

--

So we need one more change on this PR, we need to update [defaultRootKeys](https://github.com/fleetdm/fleet/blob/289dc97ec5fbe4a9568f2a1ce69556992392176f/orbit/pkg/update/update.go#L34) to have contents of https://tuf.fleetctl.com/root.json (i.e. not just the keys but whole `root.json`).\

Makes sense?

Reply:
Thank you for that analysis. I updated the `defaultRootKeys`.

I agree that since we are shipping the full metadata in the generated packages, the additional changes I made are probably unnecessary. Do you think we should leave this and change the comment to indicate it's just a fallback in case the metadata gets deleted? I suppose maybe it's best to just fail updates if that metadata gets deleted, since the default roots won't work for someone running their own update server.

Seems this also means my changelog note "new Orbit packages MUST be generated with fleetctl >= 4.21" is not necessary since older packages already do ship the full metadata?